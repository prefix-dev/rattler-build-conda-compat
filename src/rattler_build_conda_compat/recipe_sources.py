from __future__ import annotations

import sys
import typing
from typing import Any, List, TypedDict, Union

from .conditional_list import ConditionalList, visit_conditional_list

if sys.version_info < (3, 11):
    from typing_extensions import NotRequired
else:
    from typing import NotRequired

if typing.TYPE_CHECKING:
    from collections.abc import Iterator, Mapping

OptionalUrlList = Union[str, List[str], None]


class Source(TypedDict):
    url: NotRequired[str]


def get_all_url_sources(recipe: Mapping[Any, Any]) -> Iterator[str]:
    """
    Get all url sources from the recipe. This can be from a list of sources,
    a single source, or conditional and its branches.

    Arguments
    ---------
    * `recipe` - The recipe to inspect. This should be a yaml object.

    Returns
    -------
    A list of sources.
    """

    sources = recipe.get("source", None)
    sources = typing.cast(ConditionalList[Source], sources)

    # Try getting all url top-level sources
    if sources is not None:
        source_list = visit_conditional_list(sources, None)
        for source in source_list:
            if url := source.get("url"):
                yield url

    outputs = recipe.get("outputs", None)
    if outputs is None:
        return

    outputs = visit_conditional_list(outputs, None)
    for output in outputs:
        sources = output.get("source", None)
        sources = typing.cast(ConditionalList[Source], sources)
        if sources is None:
            continue
        source_list = visit_conditional_list(sources, None)
        for source in source_list:
            if url := source.get("url"):
                yield url
