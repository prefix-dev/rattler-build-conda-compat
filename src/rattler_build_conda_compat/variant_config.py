
from itertools import product
from typing import List, Mapping

def variant_combinations(data) -> List[Mapping[str, str]]:
    zip_keys = data.pop("zip_keys", [])
    # Separate the keys that need to be zipped from the rest
    zip_keys_flat = [item for sublist in zip_keys for item in sublist]
    other_keys = [key for key in data.keys() if key not in zip_keys_flat]

    # Create combinations for non-zipped keys
    other_combinations = list(product(*[data[key] for key in other_keys]))

    # Create zipped combinations
    zipped_combinations = [list(zip(*[data[key] for key in zip_group])) for zip_group in zip_keys]

    # Combine zipped combinations
    zipped_product = list(product(*zipped_combinations))

    # Combine all results into dictionaries
    final_combinations = []
    for other_combo in other_combinations:
        for zipped_combo in zipped_product:
            combined = {}
            # Add non-zipped items
            for key, value in zip(other_keys, other_combo):
                combined[key] = str(value)
            # Add zipped items
            for zip_group, zip_values in zip(zip_keys, zipped_combo):
                for key, value in zip(zip_group, zip_values):
                    combined[key] = str(value)
            final_combinations.append(combined)

    return final_combinations

