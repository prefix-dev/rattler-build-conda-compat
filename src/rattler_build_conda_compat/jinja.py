from __future__ import annotations

from typing import Any, TypedDict

import jinja2
import yaml
from jinja2 import DebugUndefined

from rattler_build_conda_compat.loader import load_yaml


class RecipeWithContext(TypedDict, total=False):
    context: dict[str, str]


class _MissingUndefined(DebugUndefined):
    def __str__(self) -> str:
        """
        By default, `DebugUndefined` return values in the form `{{ value }}`.
        `rattler-build` has a different syntax, so we need to override this method,
        and return the value in the form `${{ value }}`.
        """
        return f"${super().__str__()}"


def jinja_env() -> jinja2.Environment:
    """
    Create a `rattler-build` specific Jinja2 environment with modified syntax.
    """
    return jinja2.Environment(
        variable_start_string="${{",
        variable_end_string="}}",
        trim_blocks=True,
        lstrip_blocks=True,
        autoescape=True,
        undefined=_MissingUndefined,
    )


def load_recipe_context(context: dict[str, str], jinja_env: jinja2.Environment) -> dict[str, str]:
    """
    Load all string values from the context dictionary as Jinja2 templates.
    """
    # Process each key-value pair in the dictionary
    for key, value in context.items():
        # If the value is a string, render it as a template
        if isinstance(value, str):
            template = jinja_env.from_string(value)
            rendered_value = template.render(context)
            context[key] = rendered_value

    return context


def render_recipe_with_context(recipe_content: RecipeWithContext) -> dict[str, Any]:
    """
    Render the recipe using known values from context section.
    Unknown values are not evaluated and are kept as it is.

    Examples:
    ---
    ```python
    >>> from pathlib import Path
    >>> from rattler_build_conda_compat.loader import load_yaml
    >>> recipe_content = load_yaml((Path().resolve() / "tests" / "data" / "eval_recipe_using_context.yaml").read_text())
    >>> evaluated_context = render_recipe_with_context(recipe_content)
    >>> assert "my_value-${{ not_present_value }}" == evaluated_context["build"]["string"]
    >>>
    ```
    """
    env = jinja_env()
    context = recipe_content.get("context", {})
    # load all context templates
    context_templates = load_recipe_context(context, env)

    # render the rest of the document with the values from the context
    # and keep undefined expressions _as is_.
    template = env.from_string(yaml.dump(recipe_content))
    rendered_content = template.render(context_templates)
    return load_yaml(rendered_content)
