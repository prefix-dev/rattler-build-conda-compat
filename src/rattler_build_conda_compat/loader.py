from __future__ import annotations

from contextlib import contextmanager
from typing import TYPE_CHECKING, Any

import yaml

from rattler_build_conda_compat.conditional_list import visit_conditional_list
from rattler_build_conda_compat.utils import flatten_lists, remove_empty_keys

if TYPE_CHECKING:
    from collections.abc import Iterator
    from os import PathLike

SELECTOR_OPERATORS = ("and", "or", "not")


class RecipeLoader(yaml.BaseLoader):
    _namespace: dict[str, Any] | None = None
    _allow_missing_selector: bool = False

    @classmethod
    @contextmanager
    def with_namespace(
        cls: type[RecipeLoader],
        namespace: dict[str, Any] | None,
        *,
        allow_missing_selector: bool = False,
    ) -> Iterator[None]:
        try:
            cls._namespace = namespace
            cls._allow_missing_selector = allow_missing_selector
            yield
        finally:
            del cls._namespace

    def construct_sequence(  # noqa: C901, PLR0912
        self,
        node: yaml.ScalarNode | yaml.SequenceNode | yaml.MappingNode,
        deep: bool = False,  # noqa: FBT002, FBT001,
    ) -> list[yaml.ScalarNode]:
        """deep is True when creating an object/mapping recursively,
        in that case want the underlying elements available during construction
        """
        # find if then else selectors
        for sequence_idx, child_node in enumerate(node.value[:]):
            # if then is only present in MappingNode

            if isinstance(child_node, yaml.MappingNode):
                # iterate to find if there is IF first

                the_evaluated_one = None
                for idx, (key_node, value_node) in enumerate(child_node.value):
                    if key_node.value == "if":
                        # we catch the first one, let's try to find next pair of (then | else)
                        then_node_key, then_node_value = child_node.value[idx + 1]

                        if then_node_key.value != "then":
                            msg = "cannot have if without then, please reformat your variant file"
                            raise ValueError(msg)

                        try:
                            _, else_node_value = child_node.value[idx + 2]
                        except IndexError:
                            _, else_node_value = None, None

                        to_be_eval = f"{value_node.value}"

                        if self._allow_missing_selector:
                            split_selectors = [
                                selector
                                for selector in to_be_eval.split()
                                if selector not in SELECTOR_OPERATORS
                            ]
                            for selector in split_selectors:
                                if self._namespace and selector not in self._namespace:
                                    cleaned_selector = selector.strip("(").rstrip(")")
                                    self._namespace[cleaned_selector] = True

                        evaled = eval(to_be_eval, self._namespace)  # noqa: S307
                        if evaled:
                            the_evaluated_one = then_node_value
                        elif else_node_value:
                            the_evaluated_one = else_node_value

                        if the_evaluated_one:
                            node.value.remove(child_node)
                            node.value.insert(sequence_idx, the_evaluated_one)
                        else:
                            # neither the evaluation or else node is present, so we remove this if
                            node.value.remove(child_node)

        if not isinstance(node, yaml.SequenceNode):
            raise TypeError(
                None,
                None,
                f"expected a sequence node, but found {node.id!s}",
                node.start_mark,
            )

        return [self.construct_object(child, deep=deep) for child in node.value]


def load_yaml(content: str | bytes) -> Any:  # noqa: ANN401
    return yaml.load(content, Loader=yaml.BaseLoader)  # noqa: S506


def parse_recipe_config_file(
    path: PathLike[str], namespace: dict[str, Any] | None, *, allow_missing_selector: bool = False
) -> dict[str, Any]:
    with open(path) as f, RecipeLoader.with_namespace(
        namespace, allow_missing_selector=allow_missing_selector
    ):
        content = yaml.load(f, Loader=RecipeLoader)  # noqa: S506
    return flatten_lists(remove_empty_keys(content))


def load_all_requirements(content: dict[str, Any]) -> dict[str, Any]:
    requirements_section = dict(content.get("requirements", {}))
    if not requirements_section:
        return {}

    for section in requirements_section:
        section_reqs = requirements_section[section]
        if not section_reqs:
            continue

        requirements_section[section] = list(visit_conditional_list(section_reqs))

    return requirements_section
