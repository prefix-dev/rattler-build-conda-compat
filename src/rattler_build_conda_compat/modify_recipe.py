from __future__ import annotations

import copy
import hashlib
import io
import logging
import re
from typing import TYPE_CHECKING, Any, Literal

import requests
from ruamel.yaml import YAML

from rattler_build_conda_compat.jinja.jinja import jinja_env, load_recipe_context
from rattler_build_conda_compat.recipe_sources import Source, get_all_sources

if TYPE_CHECKING:
    from pathlib import Path

logger = logging.getLogger(__name__)

yaml = YAML()
yaml.preserve_quotes = True
yaml.width = 4096
yaml.indent(mapping=2, sequence=4, offset=2)


def _update_build_number_in_context(recipe: dict[str, Any], new_build_number: int) -> bool:
    for key in recipe.get("context", {}):
        if key.startswith("build_") or key == "build":
            recipe["context"][key] = new_build_number
            return True
    return False


def _update_build_number_in_recipe(recipe: dict[str, Any], new_build_number: int) -> bool:
    is_modified = False
    if "build" in recipe and "number" in recipe["build"]:
        recipe["build"]["number"] = new_build_number
        is_modified = True

    if "outputs" in recipe:
        for output in recipe["outputs"]:
            if "build" in output and "number" in output["build"]:
                output["build"]["number"] = new_build_number
                is_modified = True

    return is_modified


def update_build_number(file: Path, new_build_number: int = 0) -> str:
    """
    Update the build number in the recipe file.

    Arguments:
    ----------
    * `file` - The path to the recipe file.
    * `new_build_number` - The new build number to use. (default: 0)

    Returns:
    --------
    The updated recipe as a string.
    """
    with file.open("r") as f:
        data = yaml.load(f)
    build_number_modified = _update_build_number_in_context(data, new_build_number)
    if not build_number_modified:
        _update_build_number_in_recipe(data, new_build_number)

    with io.StringIO() as f:
        yaml.dump(data, f)
        return f.getvalue()


class CouldNotUpdateVersionError(Exception):
    NO_CONTEXT = "Could not find context in recipe"
    NO_VERSION = "Could not find version in recipe context"

    def __init__(self, message: str = "Could not update version") -> None:
        self.message = message
        super().__init__(self.message)


class Hash:
    def __init__(self, hash_type: Literal["md5", "sha256"], hash_value: str) -> None:
        self.hash_type = hash_type
        self.hash_value = hash_value

    def __str__(self) -> str:
        return f"{self.hash_type}: {self.hash_value}"


def _has_jinja_version(url: str) -> bool:
    """Check if the URL has a jinja `${{ version }}` in it."""
    pattern = r"\${{\s*version"
    return re.search(pattern, url) is not None


def update_hash(source: Source, url: str, hash_: Hash | None) -> None:
    """
    Update the sha256 hash in the source dictionary.

    Arguments:
    ----------
    * `source` - The source dictionary to update.
    * `url` - The URL to download and hash (if no hash is provided).
    * `hash_` - The hash to use. If not provided, the file will be downloaded and `sha256` hashed.
    """
    hash_type = hash_.hash_type if hash_ is not None else "sha256"
    # delete all old hashes that we are not updating
    all_hash_types = {"md5", "sha256"}
    for key in all_hash_types - {hash_type}:
        if key in source:
            del source[key]

    if hash_ is not None:
        source[hash_.hash_type] = hash_.hash_value
    else:
        # download and hash the file
        hasher = hashlib.sha256()
        print(f"Retrieving and hashing {url}")
        with requests.get(url, stream=True, timeout=100) as r:
            for chunk in r.iter_content(chunk_size=4096):
                hasher.update(chunk)
        source["sha256"] = hasher.hexdigest()


def update_version(file: Path, new_version: str, hash_: Hash | None) -> str:
    """
    Update the version in the recipe file.

    Arguments:
    ----------
    * `file` - The path to the recipe file.
    * `new_version` - The new version to use.
    * `hash_type` - The hash type to use. If not provided, the file will be downloaded and `sha256` hashed.

    Returns:
    --------
    The updated recipe as a string.
    """

    with file.open("r") as f:
        data = yaml.load(f)

    if "context" not in data:
        raise CouldNotUpdateVersionError(CouldNotUpdateVersionError.NO_CONTEXT)
    if "version" not in data["context"]:
        raise CouldNotUpdateVersionError(CouldNotUpdateVersionError.NO_VERSION)

    data["context"]["version"] = new_version

    # set up the jinja context
    env = jinja_env()
    context = copy.deepcopy(data.get("context", {}))
    context_variables = load_recipe_context(context, env)
    # for r-recipes we add the default `cran_mirror` variable
    context_variables["cran_mirror"] = "https://cran.r-project.org"

    for source in get_all_sources(data):
        # render the whole URL and find the hash
        if "url" not in source:
            continue

        url = source["url"]
        if isinstance(url, list):
            url = url[0]

        if not _has_jinja_version(url):
            continue

        template = env.from_string(url)
        rendered_url = template.render(context_variables)

        update_hash(source, rendered_url, hash_)

    with io.StringIO() as f:
        yaml.dump(data, f)
        return f.getvalue()
