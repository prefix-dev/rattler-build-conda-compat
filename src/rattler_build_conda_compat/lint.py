# mypy: ignore-errors

import re

from inspect import cleandoc
import os.path
import tomli
import github
import ruamel.yaml
from typing import Any, Dict, Mapping, Sequence, List
import requests
from conda.models.version import VersionOrder
from functools import lru_cache
from jsonschema import Draft202012Validator
from jsonschema import ValidationError
from textwrap import indent

from rattler_build_conda_compat.loader import load_yaml

SCHEMA_URL = "https://raw.githubusercontent.com/prefix-dev/recipe-format/main/schema.json"

REQUIREMENTS_ORDER = ["build", "host", "run"]


JINJA_VAR_PAT = re.compile(r"\${{(.*?)}}")


def _format_validation_msg(error: ValidationError):
    return cleandoc(
        f"""
        In recipe.yaml: \n{indent(error.message, " " * 12 + "> ")}
        """
    )


@lru_cache
def get_recipe_schema() -> Dict[Any, Any]:
    return requests.get(SCHEMA_URL).json()


def yaml_reader():
    # define global yaml API
    # roundrip-loader and allowing duplicate keys
    # for handling # [filter] / # [not filter]
    # Don't use a global variable for this as a global
    # variable will make conda-smithy thread unsafe.
    yaml = ruamel.yaml.YAML(typ="rt")
    yaml.allow_duplicate_keys = True
    return yaml


def lint_recipe_yaml_by_schema(recipe_file):
    schema = get_recipe_schema()
    yaml = yaml_reader()

    with open(recipe_file) as fh:
        meta = yaml.load(fh)

    validator = Draft202012Validator(schema)
    lints = []

    for error in validator.iter_errors(meta):
        lints.append(_format_validation_msg(error))
    return lints


def lint_about_contents(about_section, lints):
    for about_item in ["homepage", "license", "summary"]:
        # if the section doesn't exist, or is just empty, lint it.
        if not about_section.get(about_item, ""):
            lints.append("The {} item is expected in the about section." "".format(about_item))


def lint_recipe_maintainers(maintainers_section, lints):
    if not maintainers_section:
        lints.append(
            "The recipe could do with some maintainers listed in "
            "the `extra/recipe-maintainers` section."
        )

    if not (isinstance(maintainers_section, Sequence) and not isinstance(maintainers_section, str)):
        lints.append("Recipe maintainers should be a json list.")


def lint_recipe_tests(test_section=dict(), outputs_section=list()):
    TEST_KEYS = {"script", "python"}
    lints = []
    hints = []

    if not any(key in TEST_KEYS for key in test_section):
        if not outputs_section:
            lints.append("The recipe must have some tests.")
        else:
            has_outputs_test = False
            no_test_hints = []
            for section in outputs_section:
                test_section = section.get("tests", {})
                if any(key in TEST_KEYS for key in test_section):
                    has_outputs_test = True
                else:
                    no_test_hints.append(
                        "It looks like the '{}' output doesn't " "have any tests.".format(
                            section.get("name", "???")
                        )
                    )
            if has_outputs_test:
                hints.extend(no_test_hints)
            else:
                lints.append("The recipe must have some tests.")

    return lints, hints


def lint_license_not_unknown(license: str, lints: List):
    license = license.lower()
    if "unknown" == license.strip():
        lints.append("The recipe license cannot be unknown.")


def lint_build_number(build_section: Dict, lints: List):
    build_number = build_section.get("number", None)
    if build_number is None:
        lints.append("The recipe must have a `build/number` section.")


def lint_requirements_order(requirements_section: Dict, lints: List):
    seen_requirements = [k for k in requirements_section if k in REQUIREMENTS_ORDER]
    requirements_order_sorted = sorted(seen_requirements, key=REQUIREMENTS_ORDER.index)
    if seen_requirements != requirements_order_sorted:
        lints.append(
            "The `requirements/` sections should be defined "
            "in the following order: "
            + ", ".join(REQUIREMENTS_ORDER)
            + "; instead saw: "
            + ", ".join(seen_requirements)
            + "."
        )


def lint_package_version(package_section: dict, context_section: dict):
    package_ver = str(package_section.get("version"))
    context_ver = str(context_section.get("version"))
    ver = (
        package_ver if package_ver is not None and not package_ver.startswith("$") else context_ver
    )

    try:
        VersionOrder(ver)

    except Exception:
        return "Package version {} doesn't match conda spec".format(ver)


def lint_files_have_hash(sources_section: list, lints: list):
    for source_section in sources_section:
        if "url" in source_section and not ({"sha1", "sha256", "md5"} & set(source_section.keys())):
            lints.append(
                "When defining a source/url please add a sha256, sha1 "
                "or md5 checksum (sha256 preferably)."
            )


def lint_legacy_compilers(build_reqs):
    if build_reqs and ("toolchain" in build_reqs):
        return """Using toolchain directly in this manner is deprecated. Consider
            using the compilers outlined
            [here](https://conda-forge.org/docs/maintainer/knowledge_base.html#compilers)."""


def lint_has_recipe_file(about_section, lints):
    license_file = about_section.get("license_file", None)
    if not license_file:
        lints.append("license_file entry is missing, but is required.")


def lint_package_name(package_section: dict, context_section: dict):
    package_name = str(package_section.get("name"))
    context_name = str(context_section.get("name"))
    actual_name = (
        package_name
        if package_name is not None and not package_name.startswith("$")
        else context_name
    )

    actual_name = actual_name.strip()
    if re.match(r"^[a-z0-9_\-.]+$", actual_name) is None:
        return """Recipe name has invalid characters. only lowercase alpha, numeric, underscores, hyphens and dots allowed"""


def lint_legacy_patterns(requirements_section):
    lints = []
    build_reqs = requirements_section.get("build", None)
    if build_reqs and ("numpy x.x" in build_reqs):
        lints.append(
            "Using pinned numpy packages is a deprecated pattern.  Consider "
            "using the method outlined "
            "[here](https://conda-forge.org/docs/maintainer/knowledge_base.html#linking-numpy)."
        )
    return lints


def lint_usage_of_selectors_for_noarch(noarch_value, build_section, requirements_section):
    lints = []
    for section in requirements_section:
        section_requirements = requirements_section[section]

        if not section_requirements:
            continue

        if any(isinstance(req, dict) for req in section_requirements):
            lints.append(
                "`noarch` packages can't have skips with selectors. If "
                "the selectors are necessary, please remove "
                "`noarch: {}`.".format(noarch_value)
            )
            break

    if "skip" in build_section:
        lints.append(
            "`noarch` packages can't have skips with selectors. If "
            "the selectors are necessary, please remove "
            "`noarch: {}`.".format(noarch_value)
        )

    return lints


def lint_usage_of_single_space_in_pinned_requirements(requirements_section: dict):
    def verify_requirement(requirement, section):
        lints = []
        if "${{" in requirement:
            return lints
        parts = requirement.split()
        if len(parts) > 2 and parts[1] in [
            "!=",
            "=",
            "==",
            ">",
            "<",
            "<=",
            ">=",
        ]:
            # check for too many spaces
            lints.append(
                (
                    "``requirements: {section}: {requirement}`` should not "
                    "contain a space between relational operator and the version, i.e. "
                    "``{name} {pin}``"
                ).format(
                    section=section,
                    requirement=requirement,
                    name=parts[0],
                    pin="".join(parts[1:]),
                )
            )
            return lints
        # check that there is a space if there is a pin
        bad_char_idx = [(parts[0].find(c), c) for c in "><="]
        bad_char_idx = [bci for bci in bad_char_idx if bci[0] >= 0]
        if bad_char_idx:
            bad_char_idx.sort()
            i = bad_char_idx[0][0]
            lints.append(
                (
                    "``requirements: {section}: {requirement}`` must "
                    "contain a space between the name and the pin, i.e. "
                    "``{name} {pin}``"
                ).format(
                    section=section,
                    requirement=requirement,
                    name=parts[0][:i],
                    pin=parts[0][i:] + "".join(parts[1:]),
                )
            )

        return lints

    lints = []
    for section, requirements in requirements_section.items():
        if not requirements:
            continue
        for req in requirements:
            lints.extend(verify_requirement(req, section))
    return lints


def lint_non_noarch_dont_constrain_python_and_rbase(requirements_section):
    check_languages = ["python", "r-base"]
    host_reqs = requirements_section.get("host") or []
    run_reqs = requirements_section.get("run") or []

    lints = []

    for language in check_languages:
        filtered_host_reqs = [req for req in host_reqs if req.startswith(f"{language}")]
        filtered_run_reqs = [req for req in run_reqs if req.startswith(f"{language}")]

        if filtered_host_reqs and not filtered_run_reqs:
            lints.append(f"If {language} is a host requirement, it should be a run requirement.")

        for reqs in [filtered_host_reqs, filtered_run_reqs]:
            if language not in reqs:
                for req in reqs:
                    split = req.split(" ", 1)
                    if len(split) > 1:
                        constraint = req.split(" ", 1)[1]
                        if constraint.startswith(">") or constraint.startswith("<"):
                            lints.append(
                                f"Non noarch packages should have {language} requirement without any version constraints."
                            )

    return lints


def lint_variable_reference_should_have_space(recipe_dir, recipe_file):
    hints = []
    if recipe_dir is not None and os.path.exists(recipe_file):
        bad_vars = []
        bad_lines = []
        with open(recipe_file, "r") as fh:
            for i, line in enumerate(fh.readlines()):
                for m in JINJA_VAR_PAT.finditer(line):
                    if m.group(1) is not None:
                        var = m.group(1)
                        if var != " %s " % var.strip():
                            bad_vars.append(m.group(1).strip())
                            bad_lines.append(i + 1)
        if bad_vars:
            hints.append(
                "Jinja2 variable references are suggested to "
                "take a ``${{<one space><variable name><one space>}}``"
                " form. See lines %s." % (bad_lines,)
            )

    return hints


def lint_lower_bound_on_python(run_requirements, outputs_section):
    lints = []
    # if noarch_value == "python" and not outputs_section:
    for req in run_requirements:
        if (req.strip().split()[0] == "python") and (req != "python"):
            break
    else:
        lints.append(
            "noarch: python recipes are required to have a lower bound "
            "on the python version. Typically this means putting "
            "`python >=3.6` in **both** `host` and `run` but you should check "
            "upstream for the package's Python compatibility."
        )


def hint_pip_usage(build_section):
    hints = []

    if "script" in build_section:
        scripts = build_section["script"]
        if isinstance(scripts, str):
            scripts = [scripts]
        for script in scripts:
            if "python setup.py install" in script:
                hints.append(
                    "Whenever possible python packages should use pip. "
                    "See https://conda-forge.org/docs/maintainer/adding_pkgs.html#use-pip"
                )
    return hints


def hint_noarch_usage(build_section, requirement_section: dict):
    build_reqs = requirement_section.get("build", None)
    hints = []
    if (
        # move outside the call
        # noarch_value is None
        build_reqs
        and not any(
            [
                b.startswith("${{") and ("compiler('c')" in b or 'compiler("c")' in b)
                for b in build_reqs
            ]
        )
        and ("pip" in build_reqs)
        # move outside the call
        # and (is_staged_recipes or not conda_forge)
    ):
        no_arch_possible = True
        if "skip" in build_section:
            no_arch_possible = False

        for _, section_requirements in requirement_section.items():
            if any(isinstance(requirement, dict) for requirement in section_requirements):
                no_arch_possible = False
                break

        if no_arch_possible:
            hints.append(
                "Whenever possible python packages should use noarch. "
                "See https://conda-forge.org/docs/maintainer/knowledge_base.html#noarch-builds"
            )

    return hints


def run_conda_forge_specific(
    recipe_dir,
    package_section,
    extra_section,
    sources_section,
    requirements_section,
    outputs_section,
):
    lints = []
    hints = []

    gh = github.Github(os.environ["GH_TOKEN"])

    # Fetch list of recipe maintainers
    maintainers = extra_section.get("recipe-maintainers", [])

    recipe_dirname = os.path.basename(recipe_dir) if recipe_dir else "recipe"
    recipe_name = package_section.get("name", "").strip()
    is_staged_recipes = recipe_dirname != "recipe"

    # 1: Check that the recipe does not exist in conda-forge or bioconda
    if is_staged_recipes and recipe_name:
        cf = gh.get_user(os.getenv("GH_ORG", "conda-forge"))

        for name in set(
            [
                recipe_name,
                recipe_name.replace("-", "_"),
                recipe_name.replace("_", "-"),
            ]
        ):
            try:
                if cf.get_repo("{}-feedstock".format(name)):
                    existing_recipe_name = name
                    feedstock_exists = True
                    break
                else:
                    feedstock_exists = False
            except github.UnknownObjectException:
                feedstock_exists = False

        if feedstock_exists and existing_recipe_name == recipe_name:
            lints.append("Feedstock with the same name exists in conda-forge.")
        elif feedstock_exists:
            hints.append(
                "Feedstock with the name {} exists in conda-forge. Is it the same as this package ({})?".format(
                    existing_recipe_name,
                    recipe_name,
                )
            )

        bio = gh.get_user("bioconda").get_repo("bioconda-recipes")
        try:
            bio.get_dir_contents("recipes/{}".format(recipe_name))
        except github.UnknownObjectException:
            pass
        else:
            hints.append(
                "Recipe with the same name exists in bioconda: "
                "please discuss with @conda-forge/bioconda-recipes."
            )

        url = None
        if isinstance(sources_section, dict):
            if str(sources_section.get("url")).startswith("https://pypi.io/packages/source/"):
                url = sources_section["url"]
        else:
            for source_section in sources_section:
                if str(source_section.get("url")).startswith("https://pypi.io/packages/source/"):
                    url = source_section["url"]

        if url:
            # get pypi name from  urls like "https://pypi.io/packages/source/b/build/build-0.4.0.tar.gz"
            pypi_name = url.split("/")[6]
            mapping_request = requests.get(
                "https://raw.githubusercontent.com/regro/cf-graph-countyfair/master/mappings/pypi/name_mapping.yaml"
            )
            if mapping_request.status_code == 200:
                mapping_raw_yaml = mapping_request.content
                mapping = load_yaml(mapping_raw_yaml)
                for pkg in mapping:
                    if pkg.get("pypi_name", "") == pypi_name:
                        conda_name = pkg["conda_name"]
                        hints.append(
                            f"A conda package with same name ({conda_name}) already exists."
                        )

    # 2: Check that the recipe maintainers exists:
    for maintainer in maintainers:
        if "/" in maintainer:
            # It's a team. Checking for existence is expensive. Skip for now
            continue
        try:
            gh.get_user(maintainer)
        except github.UnknownObjectException:
            lints.append('Recipe maintainer "{}" does not exist'.format(maintainer))

    # 3: if the recipe dir is inside the example dir
    if recipe_dir is not None and "recipes/example/" in recipe_dir:
        lints.append("Please move the recipe out of the example dir and " "into its own dir.")

    # 4: Do not delete example recipe
    if is_staged_recipes and recipe_dir is not None:
        example_meta_fname = os.path.abspath(os.path.join(recipe_dir, "..", "example", "meta.yaml"))

        if not os.path.exists(example_meta_fname):
            msg = "Please do not delete the example recipe found in " "`recipes/example/meta.yaml`."

            if msg not in lints:
                lints.append(msg)

    # 5: Package-specific hints
    # (e.g. do not depend on matplotlib, only matplotlib-base)
    # TODO: do the same for if selectors
    build_reqs = requirements_section.get("build") or []
    host_reqs = requirements_section.get("host") or []
    run_reqs = requirements_section.get("run") or []
    for out in outputs_section:
        _req = out.get("requirements") or {}
        if isinstance(_req, Mapping):
            build_reqs += _req.get("build") or []
            host_reqs += _req.get("host") or []
            run_reqs += _req.get("run") or []
        else:
            run_reqs += _req

    hints_toml_url = "https://raw.githubusercontent.com/conda-forge/conda-forge-pinning-feedstock/main/recipe/linter_hints/hints.toml"
    hints_toml_req = requests.get(hints_toml_url)
    if hints_toml_req.status_code != 200:
        # too bad, but not important enough to throw an error;
        # linter will rerun on the next commit anyway
        return
    hints_toml_str = hints_toml_req.content.decode("utf-8")
    specific_hints = tomli.loads(hints_toml_str)["hints"]

    for rq in build_reqs + host_reqs + run_reqs:
        dep = rq.split(" ")[0].strip()
        if dep in specific_hints and specific_hints[dep] not in hints:
            hints.append(specific_hints[dep])

    # 6: Check if all listed maintainers have commented:
    pr_number = os.environ.get("STAGED_RECIPES_PR_NUMBER")

    if is_staged_recipes and maintainers and pr_number:
        # Get PR details using GitHub API
        current_pr = gh.get_repo("conda-forge/staged-recipes").get_pull(int(pr_number))

        # Get PR author, issue comments, and review comments
        pr_author = current_pr.user.login
        issue_comments = current_pr.get_issue_comments()
        review_comments = current_pr.get_reviews()

        # Combine commenters from both issue comments and review comments
        commenters = {comment.user.login for comment in issue_comments}
        commenters.update({review.user.login for review in review_comments})

        # Check if all maintainers have either commented or are the PR author
        non_participating_maintainers = set()
        for maintainer in maintainers:
            if maintainer not in commenters and maintainer != pr_author:
                non_participating_maintainers.add(maintainer)

        # Add a lint message if there are any non-participating maintainers
        if non_participating_maintainers:
            lints.append(
                f"The following maintainers have not yet confirmed that they are willing to be listed here: "
                f"{', '.join(non_participating_maintainers)}. Please ask them to comment on this PR if they are."
            )

    return lints, hints
