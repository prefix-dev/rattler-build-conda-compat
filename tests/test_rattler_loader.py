from __future__ import annotations

from pathlib import Path
from typing import Any

from rattler_build_conda_compat.loader import (
    load_all_requirements,
    load_yaml,
    parse_recipe_config_file,
)


def test_load_variants(snapshot, unix_namespace: dict[str, Any]) -> None:
    variants_path = Path("tests/data/variants.yaml")

    loaded_variants = parse_recipe_config_file(str(variants_path), unix_namespace)

    assert loaded_variants == snapshot


def test_load_all_requirements() -> None:
    recipe_content = Path("tests/data/recipe_requirements.yaml").read_text()

    recipe_content = load_yaml(recipe_content)

    content = load_all_requirements(recipe_content)
    print(content)


def test_load_recipe_with_missing_selectors(snapshot) -> None:
    osx_recipe = Path("tests/data/osx_recipe.yaml")

    namespace = {"osx": True, "unix": True}

    loaded_variants = parse_recipe_config_file(
        str(osx_recipe), namespace, allow_missing_selector=True
    )

    assert loaded_variants == snapshot
