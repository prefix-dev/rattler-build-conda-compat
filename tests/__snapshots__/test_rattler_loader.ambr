# serializer version: 1
# name: test_load_all_tests
  list([
    CommentedMap({
      'script': CommentedSeq([
        'echo "Hello world"',
        'pytest ./tests',
      ]),
      'requirements': CommentedMap({
        'run': CommentedSeq([
          'pytest',
        ]),
      }),
      'files': CommentedMap({
        'source': CommentedSeq([
          'tests/',
          'test.py',
          '*.sh',
        ]),
        'recipe': CommentedSeq([
          'more_tests/*.py',
        ]),
      }),
    }),
    CommentedMap({
      'python': CommentedMap({
        'imports': CommentedSeq([
          'mypkg',
          'mypkg.subpkg',
        ]),
      }),
    }),
  ])
# ---
# name: test_load_recipe_with_missing_selectors
  dict({
    'package': dict({
      'name': 'some-osx-recipe',
      'version': '1.0.0',
    }),
    'requirements': dict({
      'build': list([
        'python',
        'ruby',
        'rust',
      ]),
    }),
  })
# ---
# name: test_load_variants
  dict({
    'numpy': list([
      1.22,
      1.22,
    ]),
    'python': list([
      '3.8.* *_cpython',
      '3.9.* *_cpython',
    ]),
    'python_impl': list([
      'cpython',
      'cpython',
    ]),
    'ruby': list([
      2.5,
    ]),
  })
# ---
